README.md
This library provide CAPTCHA implementation

To execute PHPUnit tests

1. Point to /vendor/bin dir `cd ./vendor/bin`
2. Execute PHPUnit tests: `phpunit --configuration ./../../tests/phpunit.xml`
3. To check phpunit version: `phpunit --version`


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file, in reverse chronological order by release.

---

## 1.0.0 - 01 Nov 2020

### Added

- First release

---


tests/src/unit/CaptchaTest.php
<?php

namespace FRUnitTest\Captcha;

use PHPUnit\Framework\TestCase;
use FR\Captcha\Captcha;

class CaptchaTest extends TestCase
{
    protected static $CaptchaTestAsset;
    protected static $ServiceResponse;
    protected static $CaptchaStorage;
    protected static $Captcha;
    protected static $config;

    /**
     * This method is executed only once per class
     *
     * @return void
     */
    public static function setUpBeforeClass(): void
    {
        self::$CaptchaTestAsset = new CaptchaTestAsset();

        self::$ServiceResponse = self::$CaptchaTestAsset->getServiceResponse();
        self::$CaptchaStorage = self::$CaptchaTestAsset->getCaptchaStorage();
        self::$Captcha = self::$CaptchaTestAsset->getCaptcha();
        self::$config = self::$CaptchaTestAsset->getConfig();
    }

    /**
     * @test
     * @covers FR\Captcha\Captcha::__construct
     * 
     * @return void
     */
    public function constructor()
    {
        // validate captcha_lifetime
        $invalid_captcha_lifetime = ['', '80'];
        foreach ($invalid_captcha_lifetime as $i => $captcha_lifetime) {
            $exception = false;
            try {
                // Override $config with invalid value
                $config = self::$config;
                $config['captcha_lifetime'] = $captcha_lifetime;

                new Captcha(
                    self::$ServiceResponse,
                    $config,
                    self::$CaptchaStorage
                );
            } catch (\Exception $expected) {
                $exception = true;
            }
            $this->assertTrue($exception, 'Exception not thrown: `captcha_lifetime` cannot be empty and must be integer');
        }

        // validate captcha_code_length
        $invalid_captcha_code_length = ['', '80', 2, 11];
        foreach ($invalid_captcha_code_length as $i => $captcha_code_length) {
            $exception = false;
            try {
                // Override $config with invalid value
                $config = self::$config;
                $config['captcha_code_length'] = $captcha_code_length;

                new Captcha(
                    self::$ServiceResponse,
                    $config,
                    self::$CaptchaStorage
                );
            } catch (\Exception $expected) {
                $exception = true;
            }
            $this->assertTrue($exception, 'Exception not thrown: `captcha_code_length` cannot be empty and must be integer and must be from 4 to 10 chars inclusive');
        }

        // validate captcha_image_base_path
        $invalid_captcha_image_base_path = ['', true, '/dir/not/exists', TEST_FR_CAPTCHA_BASE_PATH . '/README.md'];
        foreach ($invalid_captcha_image_base_path as $i => $captcha_image_base_path) {
            $exception = false;
            try {
                // Override $config with invalid value
                $config = self::$config;
                $config['captcha_image_base_path'] = $captcha_image_base_path;

                new Captcha(
                    self::$ServiceResponse,
                    $config,
                    self::$CaptchaStorage
                );
            } catch (\Exception $expected) {
                $exception = true;
            }

            if (in_array($i, [0, 1]))
                $this->assertTrue($exception, 'Exception not thrown: `captcha_image_base_path` cannot be empty and must be string');
            if (in_array($i, [2, 3]))
                $this->assertTrue($exception, 'Exception not thrown: invalid `captcha_image_base_path` must be dir and must exists with writable permissions');
        }

        // validate captcha_image_base_url
        $invalid_captcha_image_base_url = ['', true, 'http//invalid-url'];
        foreach ($invalid_captcha_image_base_url as $i => $captcha_image_base_url) {
            $exception = false;
            try {
                // Override $config with invalid value
                $config = self::$config;
                $config['captcha_image_base_url'] = $captcha_image_base_url;

                new Captcha(
                    self::$ServiceResponse,
                    $config,
                    self::$CaptchaStorage
                );
            } catch (\Exception $expected) {
                $exception = true;
            }
            $this->assertTrue($exception, 'Exception not thrown: `captcha_image_base_url` cannot be empty and must be string and must start with http:// or https://');
        }
    }

    /**
     * @test
     * @covers FR\Captcha\Captcha::generateUniqueId
     * 
     * @return void
     */
    public function generateUniqueId()
    {
        $test = [
            [
                'length' => 32,
            ],
            [
                'length' => '64',
            ],
            [
                'length' => 128,
            ],
            [
                'length' => '256',
            ],
        ];

        foreach ($test as $i => $param) {
            $token = invokeMethod(
                self::$Captcha,
                'generateUniqueId',
                [$param['length']]
            );

            $this->assertEquals($param['length'], strlen($token));
        }
    }

    /**
     * @test
     * @covers FR\Captcha\Captcha::createCaptcha
     * 
     * @return void
     */
    public function createCaptcha()
    {
        $response =  invokeMethod(
            self::$Captcha,
            'createCaptcha',
            []
        );

        $this->assertEquals(200, @$response['code']);
        $this->assertEquals('success', @$response['status']);
        $this->assertIsArray(@$response['data']);
        $this->assertNotEmpty(@$response['data']);
        $this->assertArrayHasKey('captcha_id', @$response['data']);
        $this->assertArrayHasKey('captcha_image_url', @$response['data']);
        $captcha_image_file_path = self::$config['captcha_image_base_path'] . '/' . basename($response['data']['captcha_image_url']);
        $this->assertFileExists($captcha_image_file_path);

        // Test created captcha must be valid
        $row = self::$CaptchaStorage->getCaptchaById(@$response['data']['captcha_id']);
        $this->assertEquals(self::$config['captcha_code_length'], strlen($row['captcha_code']));
        $this->assertTrue((time() < strtotime($row['captcha_expired_at'])));

        // Clear database once tested
        self::$CaptchaStorage->deleteCaptchaById(@$response['data']['captcha_id']);

        return $row['captcha_image'];
    }

    /**
     * @test
     * @covers FR\Captcha\Captcha::removeCaptchaImageFile
     * @depends createCaptcha
     * 
     * @return void
     */
    public function removeCaptchaImageFile($captcha_image)
    {
        invokeMethod(
            self::$Captcha,
            'removeCaptchaImageFile',
            [$captcha_image]
        );

        $captcha_image_file_path = self::$config['captcha_image_base_path'] . '/' . $captcha_image;
        $this->assertFileNotExists($captcha_image_file_path);
    }

    /**
     * @test
     * @covers FR\Captcha\Captcha::verifyCaptcha
     * 
     * @return void
     */
    public function verifyCaptcha()
    {
        // Create expired captcha
        $expired_captcha_id = generateUniqueId(32);
        $expired_captcha_code = '123456';
        self::$CaptchaStorage->insertCaptcha(
            $expired_captcha_id,
            strtolower($expired_captcha_id . '.png'),
            $expired_captcha_code,
            date('Y-m-d H:i:s', strtotime('-1 Second')),
            date('Y-m-d H:i:s')
        );

        // Create valid captcha
        $valid_captcha_id = generateUniqueId(32);
        $valid_captcha_code = 'abC123';
        self::$CaptchaStorage->insertCaptcha(
            $valid_captcha_id,
            strtolower($valid_captcha_id . '.png'),
            $valid_captcha_code,
            date('Y-m-d H:i:s', strtotime('+1 Hour')),
            date('Y-m-d H:i:s')
        );

        $test = [
            [
                'captcha_id' => ' ',
                'captcha_code' => 'not-empty'
            ],
            [
                'captcha_id' => 'not-empty',
                'captcha_code' => ' '
            ],
            [
                'captcha_id' => 'invalid-captcha-id',
                'captcha_code' => 'not-empty'
            ],
            [
                'captcha_id' => $valid_captcha_id,
                'captcha_code' => 'invalid'
            ],
            [
                'captcha_id' => $expired_captcha_id,
                'captcha_code' => '123456'
            ],
            [
                'captcha_id' => $valid_captcha_id,
                'captcha_code' => ' ABC123 '
            ],
        ];

        foreach ($test as $i => $param) {
            $response =  invokeMethod(
                self::$Captcha,
                'verifyCaptcha',
                [
                    $param['captcha_id'],
                    $param['captcha_code']
                ]
            );

            if (in_array($i, [0, 1, 2, 3, 4])) {
                $this->assertEquals(400, @$response['code']);
                $this->assertEquals('error', @$response['status']);
            }
            if (in_array($i, [0]))
                $this->assertEquals('captcha_id_required', @$response['type']);
            if (in_array($i, [1]))
                $this->assertEquals('captcha_code_required', @$response['type']);
            if (in_array($i, [2]))
                $this->assertEquals('captcha_not_found', @$response['type']);
            if (in_array($i, [3]))
                $this->assertEquals('invalid_captcha_code', @$response['type']);
            if (in_array($i, [4]))
                $this->assertEquals('expired_captcha_code', @$response['type']);
            if (in_array($i, [5])) {
                $this->assertEquals(200, @$response['code']);
                $this->assertEquals('success', @$response['status']);
            }
        }

        // Clear database once tested
        self::$CaptchaStorage->deleteCaptchaById($valid_captcha_id);
        self::$CaptchaStorage->deleteCaptchaById($expired_captcha_id);
    }
}


tests/src/unit/CaptchaTestAsset.php
<?php

namespace FRUnitTest\Captcha;

use PHPUnit\Framework\TestCase;
use FRUnitTest\Captcha\Storage\MySQL\MySQLTestAsset;
use FRUnitTest\Captcha\Storage\Oracle\OracleTestAsset;
use FR\ServiceResponse\ServiceResponse;
use FR\Captcha\Captcha;

class CaptchaTestAsset extends TestCase
{
    public function getServiceResponse()
    {
        $ServiceResponse = new ServiceResponse();
        return $ServiceResponse;
    }

    public function getConfig()
    {
        $config = [
            // 1 Hour
            'captcha_lifetime'              => 3600,
            // Length of captcha code
            'captcha_code_length'           => 6,
            // Path to save captcha images
            'captcha_image_base_path'       => TEST_FR_CAPTCHA_IMAGE_BASE_PATH,
            // Base url for captcha images
            'captcha_image_base_url'        => TEST_FR_CAPTCHA_IMAGE_BASE_URL,
        ];
        return $config;
    }

    public function getCaptchaStorage()
    {
        if (TEST_FR_CAPTCHA_STORAGE == 'Oracle') {
            if (!TEST_FR_CAPTCHA_STORAGE_ORACLE)
                throw new \Exception('TEST_FR_CAPTCHA_STORAGE_ORACLE is not enabled in phpunit.xml');

            $StorageTestAsset =  new OracleTestAsset();
        } else if (TEST_FR_CAPTCHA_STORAGE == 'MySQL') {
            if (!TEST_FR_CAPTCHA_STORAGE_MYSQL)
                throw new \Exception('TEST_FR_CAPTCHA_STORAGE_MYSQL is not enabled in phpunit.xml');

            $StorageTestAsset =  new MySQLTestAsset();
        } else {
            throw new \Exception('Invalid value for TEST_FR_CAPTCHA_STORAGE in phpunit.xml');
        }

        return $StorageTestAsset->getCaptchaStorage();
    }

    public function getCaptcha()
    {
        $Captcha = new Captcha(
            $this->getServiceResponse(),
            $this->getConfig(),
            $this->getCaptchaStorage()
        );
        return $Captcha;
    }
}


tests/src/unit/Storage/MySQL/MySQLTestAsset.php
<?php

namespace FRUnitTest\Captcha\Storage\MySQL;

use PHPUnit\Framework\TestCase;
use FR\Captcha\Storage\MySQL\CaptchaStorage;
use FR\Db\DbFactory;

class MySQLTestAsset extends TestCase
{
    public function getDB()
    {
        // Skip this test if disabled
        if (!TEST_FR_CAPTCHA_STORAGE_MYSQL)
            $this->markTestSkipped('TEST_FR_CAPTCHA_STORAGE_MYSQL is disabled in phpunit.xml');

        // MySQL connection configuration
        $config =  array(
            'driver'    => 'TEST_FR_CAPTCHA_STORAGE_MYSQL_DRIVER',
            'hostname'  => 'TEST_FR_CAPTCHA_STORAGE_MYSQL_HOSTNAME',
            'port'      => 'TEST_FR_CAPTCHA_STORAGE_MYSQL_PORT',
            'username'  => 'TEST_FR_CAPTCHA_STORAGE_MYSQL_USERNAME',
            'password'  => 'TEST_FR_CAPTCHA_STORAGE_MYSQL_PASSWORD',
            'database'  => 'TEST_FR_CAPTCHA_STORAGE_MYSQL_DATABASE',
            'charset'   => 'TEST_FR_CAPTCHA_STORAGE_MYSQL_CHARSET',
        );

        // Validate connection configurations
        foreach ($config as $key => $value) {
            if (!defined($value)) {
                throw new \Exception('const ' . $value . ' not defined in phpunit.xml');
            }

            if (empty(constant($value)) && !in_array($key, ['password'])) {
                throw new \Exception('value required for const ' . $value . ' in phpunit.xml');
            }

            // Assign constant value in $config
            $config[$key] = constant($value);
        }

        // Create $db object and connect to database
        $DB = new DbFactory();
        $DB = $DB->init($config);
        return $DB;
    }

    public function getCaptchaTableName()
    {
        return strtolower(TEST_FR_CAPTCHA_STORAGE_MYSQL_CAPTCHA_TABLE_NAME);
    }

    public function getCaptchaStorage()
    {
        return new CaptchaStorage(
            $this->getDB(),
            $this->getCaptchaTableName()
        );
    }
}


src/Storage/Oracle/CaptchaStorage.php
<?php

namespace FR\Captcha\Storage\Oracle;

use FR\Db\DbInterface;
use FR\Captcha\Storage\CaptchaStorageInterface;

class CaptchaStorage implements CaptchaStorageInterface
{
    /**
     * @var object FR\Db\DbInterface
     */
    protected $DB;

    /**
     * Captcha table name Like: schema.table_name
     *
     * @var string
     */
    protected $captcha_table_name;

    /**
     * Captcha storage
     *
     * @param object FR\Db\DbInterface $DB
     * @param string $captcha_table_name Like: schema.table_name
     * @throws \Exception `captcha_table_name` cannot be empty and must be string
     */
    public function __construct(DBInterface $DB, $captcha_table_name)
    {
        if (
            !$captcha_table_name ||
            !is_string($captcha_table_name)
        )
            throw new \Exception('`captcha_table_name` cannot be empty and must be string');

        $parts = explode('.', $captcha_table_name);
        if (count($parts) != 2)
            throw new \Exception('`captcha_table_name` name format must be like: schema.table_name');

        $this->DB = $DB;
        $this->captcha_table_name = strtoupper($captcha_table_name);
    }

    /**
     * Return SQL script of database structure
     *
     * @return string
     */
    public function getDBStructure()
    {
        $parts   = explode('.', $this->captcha_table_name);
        @$schema = $parts[0];
        @$table  = $parts[1];

        $script = " CREATE TABLE " . $this->captcha_table_name . "
                    (
                      CAPTCHA_ID          RAW(16)                  NOT NULL,
                      CAPTCHA_IMAGE       VARCHAR2(100 CHAR)       NOT NULL,
                      CAPTCHA_CODE        VARCHAR2(10 CHAR)        NOT NULL,
                      CAPTCHA_EXPIRED_AT  DATE                     NOT NULL,
                      CAPTCHA_CREATED_AT  DATE                     NOT NULL
                    );
                    CREATE UNIQUE INDEX " . $this->captcha_table_name . "_PK ON " . $this->captcha_table_name . " (CAPTCHA_ID);
                    ALTER TABLE " . $this->captcha_table_name . " ADD (
                       CONSTRAINT " . $table . "_PK
                       PRIMARY KEY
                       (CAPTCHA_ID)
                       USING INDEX " . $this->captcha_table_name . "_PK
                       ENABLE VALIDATE 
                    ); ";

        return $script;
    }

    /**
     * Create database structure if already not created
     *
     * @return bool return true when created otherwise false
     */
    public function createDBStructure()
    {
        $query = ' SELECT OWNER, 
                          TABLE_NAME 
                    FROM ALL_TABLES
                        WHERE UPPER(OWNER || \'.\' || TABLE_NAME)
                                IN (:CAPTCHA_TABLE_NAME) ';
        $values = [
            ':CAPTCHA_TABLE_NAME' => str_replace('"', '', strtoupper($this->captcha_table_name)),
        ];
        $tables = $this->DB->fetchColumn($query, $values);
        if (empty($tables)) {
            $query = $this->getDBStructure();
            $this->DB->importSQL($query);
            return true;
        }

        return false;
    }

    /**
     * Get captcha details by captcha_id
     * captcha_id is case insensitive
     *
     * @param string $captcha_id
     * @return array
     */
    public function getCaptchaById($captcha_id)
    {
        $captcha_id = strtolower($captcha_id);
        $exp = $this->DB->getExpression();

        $query
            = " SELECT " . $exp->getUuid('CAPTCHA_ID') . " CAPTCHA_ID,
                            CAPTCHA_IMAGE,
                            CAPTCHA_CODE,
                        " . $exp->getDate("CAPTCHA_EXPIRED_AT") . " CAPTCHA_EXPIRED_AT,
                        " . $exp->getDate("CAPTCHA_CREATED_AT") . " CAPTCHA_CREATED_AT
                FROM    " . $this->captcha_table_name . "
                WHERE   " . $exp->getUuid('CAPTCHA_ID') . " = :CAPTCHA_ID ";
        $values = [
            ':CAPTCHA_ID' => $captcha_id
        ];
        $row = $this->DB->fetchRow($query, $values);
        return $row;
    }

    /**
     * Get all expired captchas
     *
     * @return array
     */
    public function getExpiredCaptchas()
    {
        $exp = $this->DB->getExpression();

        $query
            = " SELECT " . $exp->getUuid('CAPTCHA_ID') . " CAPTCHA_ID,
                            CAPTCHA_IMAGE,
                            CAPTCHA_CODE,
                        " . $exp->getDate("CAPTCHA_EXPIRED_AT") . " CAPTCHA_EXPIRED_AT,
                        " . $exp->getDate("CAPTCHA_CREATED_AT") . " CAPTCHA_CREATED_AT
                FROM    " . $this->captcha_table_name . "
                WHERE   " . $exp->getDate("CAPTCHA_EXPIRED_AT") . " <= :CAPTCHA_EXPIRED_AT ";
        $values = [
            ':CAPTCHA_EXPIRED_AT' => date('Y-m-d H:i:s')
        ];
        $rows = $this->DB->fetchRows($query, $values);
        return $rows;
    }

    /**
     * Insert captcha details
     *
     * @param string $captcha_id
     * @param string $captcha_image 
     * @param string $captcha_code
     * @param string $captcha_expired_at  Datetime format: Y-m-d H:i:s
     * @param string $captcha_created_at Datetime format: Y-m-d H:i:s
     * @return void
     */
    public function insertCaptcha(
        $captcha_id,
        $captcha_image,
        $captcha_code,
        $captcha_expired_at,
        $captcha_created_at
    ) {
        if (strlen($captcha_id) != 32)
            throw new \Exception('`captcha_id` must be equal to 32 char');

        $exp = $this->DB->getExpression();

        $data = [];
        $data['CAPTCHA_ID'] = $exp->setUuid($captcha_id);
        $data['CAPTCHA_IMAGE'] = $captcha_image;
        $data['CAPTCHA_CODE'] = $captcha_code;
        $data['CAPTCHA_EXPIRED_AT'] = $exp->setDate($captcha_expired_at);
        $data['CAPTCHA_CREATED_AT'] = $exp->setDate($captcha_created_at);
        $this->DB->insert($this->captcha_table_name, $data);
    }

    /**
     * Delete captcha details by captcha_id
     *
     * @param string $captcha_id
     * @return void
     */
    public function deleteCaptchaById($captcha_id)
    {
        $captcha_id = strtolower($captcha_id);
        $exp = $this->DB->getExpression();

        $query = " DELETE FROM " . $this->captcha_table_name . "
                   WHERE " . $exp->getUuid('CAPTCHA_ID') . " = :CAPTCHA_ID ";
        $values = [
            ':CAPTCHA_ID' => $captcha_id
        ];
        $this->DB->delete($query, $values);
    }

    /**
     * Delete all expired captchas
     *
     * @return void
     */
    public function deleteExpiredCaptchas()
    {
        $exp = $this->DB->getExpression();

        $query = " DELETE FROM " . $this->captcha_table_name . "
                    WHERE   " . $exp->getDate("CAPTCHA_EXPIRED_AT") . " <= :CAPTCHA_EXPIRED_AT ";
        $values = [
            ':CAPTCHA_EXPIRED_AT' => date('Y-m-d H:i:s')
        ];
        $this->DB->delete($query, $values);
    }
}


src/Storage/CaptchaStorageInterface.php
<?php

namespace FR\Captcha\Storage;

interface CaptchaStorageInterface
{
    /**
     * Get captcha details by captcha_id
     * captcha_id is case insensitive
     *
     * @param string $captcha_id
     * @return array
     */
    public function getCaptchaById($captcha_id);

    /**
     * Get all expired captchas
     *
     * @return array
     */
    public function getExpiredCaptchas();

    /**
     * Insert captcha details
     *
     * @param string $captcha_id
     * @param string $captcha_image 
     * @param string $captcha_code
     * @param string $captcha_expired_at  Datetime format: Y-m-d H:i:s
     * @param string $captcha_created_at Datetime format: Y-m-d H:i:s
     * @return void
     */
    public function insertCaptcha(
        $captcha_id,
        $captcha_image,
        $captcha_code,
        $captcha_expired_at,
        $captcha_created_at
    );

    /**
     * Delete captcha details by captcha_id
     *
     * @param string $captcha_id
     * @return void
     */
    public function deleteCaptchaById($captcha_id);

    /**
     * Delete all expired captchas
     *
     * @return void
     */
    public function deleteExpiredCaptchas();
}


src/Captcha.php
<?php

namespace FR\Captcha;

use FR\ServiceResponse\ServiceResponseInterface;

/**
 * @author Faisal Rehman <faisalrehmanid@hotmail.com>
 * 
 * This class provide Captcha implementation
 * 
 * Example: How to use this class?
 * 
 * ```
 * <?php
 *      // Provide constructor parameters accordingly
 *      $Captcha = new \FR\Captcha\Captcha();
 * 
 *      // Create Captcha
 *      $ServiceResponse = $Captcha->createCaptcha();
 *      
 *      // Verify Captcha
 *      $ServiceResponse = $Captcha->verifyCaptcha($captcha_id, $captcha_code);
 * 
 *      // Delete all expired captchas
 *      $ServiceResponse = $Captcha->deleteExpiredCaptchas(); 
 * ?>
 * ```
 */
class Captcha
{
    protected $ServiceResponse;
    protected $config;
    protected $CaptchaStorage;

    /**
     * Create Captcha object
     *
     * @param object \FR\ServiceResponse\ServiceResponseInterface $ServiceResponse
     * @param array $config = [  // 1 Hour
     *                           'captcha_lifetime'              => 3600,
     *                           // Length of captcha code
     *                           'captcha_code_length'           => 6,
     *                           // Path to save captcha images
     *                           'captcha_image_base_path'       => '/complete/path/to/store/captcha/images',
     *                           // Base url for captcha images
     *                           'captcha_image_base_url'        => 'https://test.com/captcha/images'];
     * 
     * @param object Storage\CaptchaStorageInterface $CaptchaStorage
     */
    public function __construct(
        ServiceResponseInterface $ServiceResponse,
        array $config,
        Storage\CaptchaStorageInterface $CaptchaStorage
    ) {
        @$captcha_lifetime = $config['captcha_lifetime'];
        @$captcha_code_length = $config['captcha_code_length'];
        @$captcha_image_base_path = $config['captcha_image_base_path'];
        @$captcha_image_base_url = $config['captcha_image_base_url'];

        if (!$captcha_lifetime || !is_int($captcha_lifetime))
            throw new \Exception('`captcha_lifetime` cannot be empty and must be integer');

        if (
            !$captcha_code_length ||
            !is_int($captcha_code_length) ||
            ($captcha_code_length < 4) ||
            ($captcha_code_length > 10)
        )
            throw new \Exception('`captcha_code_length` cannot be empty and must be integer and must be from 4 to 10 chars inclusive');

        if (!$captcha_image_base_path || !is_string($captcha_image_base_path))
            throw new \Exception('`captcha_image_base_path` cannot be empty and must be string');
        if (!is_writable($captcha_image_base_path) || !is_dir($captcha_image_base_path))
            throw new \Exception('invalid `captcha_image_base_path` value is `' . $captcha_image_base_path . '` must be dir and must exists with writable permissions');

        if (
            !$captcha_image_base_url ||
            !is_string($captcha_image_base_url) ||
            (substr($captcha_image_base_url, 0, 7) != 'http://' &&
                substr($captcha_image_base_url, 0, 8) != 'https://')
        )
            throw new \Exception('`captcha_image_base_url` cannot be empty and must be string and must start with http:// or https://');

        $this->ServiceResponse = $ServiceResponse;
        $this->config = $config;
        $this->CaptchaStorage = $CaptchaStorage;
    }

    /**
     * Generate Unique ID of fixed length
     *
     * @param int $length
     * @return string
     */
    protected function generateUniqueId($length)
    {
        $length = intval($length) / 2;

        if (function_exists('random_bytes')) {
            $random = random_bytes($length);
        }

        if (function_exists('openssl_random_pseudo_bytes')) {
            $random = openssl_random_pseudo_bytes($length);
        }

        if ($random !== false && strlen($random) === $length) {
            return  bin2hex($random);
        }

        $unique_id = '';
        $characters = '0123456789abcdef';
        for ($i = 0; $i < ($length * 2); $i++) {
            $unique_id .= $characters[rand(0, strlen($characters) - 1)];
        }

        return $unique_id;
    }

    /**
     * Generate captcha image
     *
     * @param string $captcha_id
     * @param string $captcha_code
     * @return string file name
     */
    protected function generateCaptchaImage($captcha_id, $captcha_code)
    {
        $count = strlen($captcha_code);
        $image_width = 175;
        $image_height = 50;

        // Initialise image with dimensions in pixels
        $image = @imagecreatetruecolor($image_width, $image_height);

        // Set background to white
        $background = imagecolorallocate($image, 0xFF, 0xFF, 0xFF);
        imagefill($image, 0, 0, $background);

        // Draw random lines on canvas
        for ($i = 0; $i < 6; $i++) {
            imagesetthickness($image, rand(1, 3));
            $linecolors = [0xCC, 0xBB, 0xAA];
            $color = $linecolors[rand(0, 2)];
            $linecolor = imagecolorallocate($image, $color, $color, $color);
            imageline($image, 0, rand(0, $image_height), $image_width, rand(0, $image_height), $linecolor);
        }

        // Add captcha code to image
        $x = 6;
        for ($i = 0; $i < $count; $i++) {
            $fontfile = __DIR__ . '/fonts/' . rand(1, 5) . '.ttf';
            $textcolors = [0x00, 0x11, 0x22, 0x33];
            $color = $textcolors[rand(0, 3)];
            $textcolor = imagecolorallocate($image, $color, $color, $color);
            imagettftext($image, rand(18, 20), rand(-15, 15), $x, rand(30, 40), $textcolor, $fontfile, $captcha_code[$i]);
            $x += (($image_width - 12) / $count);
        }

        // Save captcha image
        ob_start();
        imagepng($image);
        imagedestroy($image);
        $data = ob_get_clean();

        $file_name = $captcha_id . '.png';
        $file_path = $this->config['captcha_image_base_path'] . '/' . $file_name;
        file_put_contents($file_path, $data);

        return $file_name;
    }

    /**
     * Remove captcha image file
     *
     * @param string $captcha_image_file_name
     * @return void
     */
    protected function removeCaptchaImageFile($captcha_image_file_name)
    {
        $file_path = $this->config['captcha_image_base_path'] . '/' . $captcha_image_file_name;
        if (is_file($file_path)) {
            unlink($file_path);
        }
    }

    /**
     * Create captcha
     * 
     * @return array of ServiceResponse
     */
    public function createCaptcha()
    {
        $captcha_id = $this->generateUniqueId(32);

        $captcha_code = '';
        $characters = '1A2a3B4b5C6c7D8d9EeFf';
        for ($i = 0; $i < $this->config['captcha_code_length']; $i++) {
            $captcha_code .= $characters[rand(0, strlen($characters) - 1)];
        }

        $captcha_image = $this->generateCaptchaImage($captcha_id, $captcha_code);
        $captcha_expired_at = date('Y-m-d H:i:s', time() + $this->config['captcha_lifetime']);
        $captcha_created_at = date('Y-m-d H:i:s');

        $this->CaptchaStorage->insertCaptcha(
            $captcha_id,
            $captcha_image,
            $captcha_code,
            $captcha_expired_at,
            $captcha_created_at
        );

        $data = [
            'captcha_id' => $captcha_id,
            'captcha_image_url' => $this->config['captcha_image_base_url'] . '/' . $captcha_image
        ];
        return $this->ServiceResponse->success(200, $data)->toArray();
    }

    /**
     * Verify captcha
     *
     * @param string $captcha_id Case insensitive
     * @param string $captcha_code Case insensitive
     * @return array of ServiceResponse
     */
    public function verifyCaptcha($captcha_id, $captcha_code)
    {
        $captcha_id = trim($captcha_id);
        $captcha_code = trim($captcha_code);

        if (!$captcha_id)
            return $this->ServiceResponse->error(400, 'captcha_id_required', 'captcha_id could not be empty')->toArray();
        if (!$captcha_code)
            return $this->ServiceResponse->error(400, 'captcha_code_required', 'captcha_code could not be empty')->toArray();

        $captcha = $this->CaptchaStorage->getCaptchaById($captcha_id);
        if (empty($captcha))
            return $this->ServiceResponse->error(400, 'captcha_not_found', 'Captcha not found')->toArray();

        if (strtolower($captcha_code) != strtolower($captcha['captcha_code']))
            return $this->ServiceResponse->error(400, 'invalid_captcha_code', 'Invalid captcha code')->toArray();

        if (time() > strtotime($captcha['captcha_expired_at']))
            return $this->ServiceResponse->error(400, 'expired_captcha_code', 'Captcha has been expired')->toArray();

        return $this->ServiceResponse->success(200)->toArray();
    }

    /**
     * Delete all expired captchas
     * 
     * @return array of ServiceResponse
     */
    public function deleteExpiredCaptchas()
    {
        $captchas = $this->CaptchaStorage->getExpiredCaptchas();
        foreach ($captchas as $i => $captcha) {
            // Delete captcha image file from server
            $this->removeCaptchaImageFile($captcha['captcha_image']);
        }

        $this->CaptchaStorage->deleteExpiredCaptchas();
        return $this->ServiceResponse->success(200)->toArray();
    }
}
